File: cpp/equity-var-engine/tests/test_backtest.cpp
// Backtesting: Kupiec LR hand-computed, chi2 p-values, Christoffersen
// clustering detection, Basel traffic-light exact zone boundaries.
#include <gtest/gtest.h>

#include <cmath>
#include <cstdint>
#include <vector>

#include "eqvar/backtest.hpp"
#include "eqvar/stats.hpp"

using namespace eqvar;

TEST(Exceptions, IndicatorAndBroadcast) {
    const std::vector<double> pnl = {-120.0, 50.0, -99.9, -100.1, 0.0};
    const std::vector<double> var1 = {100.0};  // scalar broadcast
    const std::vector<std::uint8_t> ex = exceptions_from_pnl(pnl, var1);
    const std::vector<std::uint8_t> expected = {1, 0, 0, 1, 0};
    EXPECT_EQ(ex, expected);
    const std::vector<double> vard = {130.0, 100.0, 90.0, 100.0, 100.0};
    EXPECT_EQ(exceptions_from_pnl(pnl, vard), (std::vector<std::uint8_t>{0, 0, 1, 1, 0}));
    EXPECT_THROW(exceptions_from_pnl(pnl, std::vector<double>{-1.0}), std::invalid_argument);
    EXPECT_THROW(exceptions_from_pnl(pnl, std::vector<double>{1.0, 2.0}), std::invalid_argument);
}

TEST(Kupiec, HandComputedLrTo1e12) {
    // T = 250, x = 5, p = 0.01: LR computed from first principles here.
    const double T = 250.0, x = 5.0, p = 0.01, pihat = x / T;
    const double ll0 = (T - x) * std::log(1.0 - p) + x * std::log(p);
    const double ll1 = (T - x) * std::log(1.0 - pihat) + x * std::log(pihat);
    const double lr_hand = -2.0 * (ll0 - ll1);
    const KupiecResult r = kupiec_pof(250, 5, 0.01);
    EXPECT_NEAR(r.lr, lr_hand, 1e-12);
    EXPECT_NEAR(r.pvalue, chi2_sf(lr_hand, 1.0), 1e-14);
    EXPECT_DOUBLE_EQ(r.expected, 2.5);
    EXPECT_DOUBLE_EQ(r.rate, 0.02);
}

TEST(Kupiec, Chi2CriticalValueGivesPointOhFive) {
    // Find the exception count whose LR brackets 3.8415 and confirm the
    // p-value machinery: chi2(1) sf at the critical value is 0.05 to 1e-4.
    EXPECT_NEAR(chi2_sf(3.841458820694124, 1.0), 0.05, 1e-4);
    // Exact-coverage sample: x/T == alpha -> LR = 0, p = 1.
    const KupiecResult exact = kupiec_pof(200, 2, 0.01);
    EXPECT_NEAR(exact.lr, 0.0, 1e-12);
    EXPECT_DOUBLE_EQ(exact.pvalue, 1.0);
}

TEST(Kupiec, DegenerateCountsAndMonotonicity) {
    // Zero exceptions on 250 days at 1 %: LR = -2 * 250 ln(0.99).
    const KupiecResult zero = kupiec_pof(250, 0, 0.01);
    EXPECT_NEAR(zero.lr, -2.0 * 250.0 * std::log(0.99), 1e-12);
    // More excess exceptions -> larger LR.
    EXPECT_GT(kupiec_pof(250, 10, 0.01).lr, kupiec_pof(250, 6, 0.01).lr);
    // A materially bad model is rejected at 5 %.
    EXPECT_LT(kupiec_pof(250, 10, 0.01).pvalue, 0.05);
    EXPECT_THROW(kupiec_pof(0, 0, 0.01), std::invalid_argument);
    EXPECT_THROW(kupiec_pof(250, 251, 0.01), std::invalid_argument);
    EXPECT_THROW(kupiec_pof(250, -1, 0.01), std::invalid_argument);
}

TEST(Christoffersen, TransitionCountsOnTinyPattern) {
    const std::vector<std::uint8_t> ex = {0, 1, 1, 0, 1};
    const ChristoffersenResult r = christoffersen_independence(ex);
    EXPECT_DOUBLE_EQ(r.n00, 0.0);
    EXPECT_DOUBLE_EQ(r.n01, 2.0);
    EXPECT_DOUBLE_EQ(r.n10, 1.0);
    EXPECT_DOUBLE_EQ(r.n11, 1.0);
    EXPECT_DOUBLE_EQ(r.pi01, 1.0);
    EXPECT_DOUBLE_EQ(r.pi11, 0.5);
}

TEST(Christoffersen, DetectsPlantedClustering) {
    // 250 days, 10 exceptions: (a) one solid run of 10 -> heavy clustering,
    // (b) evenly spread every 25 days -> independent-looking.
    std::vector<std::uint8_t> clustered(250, 0), spread(250, 0);
    for (int t = 100; t < 110; ++t) clustered[t] = 1;
    for (int t = 12; t < 250; t += 25) spread[t] = 1;
    const ChristoffersenResult c = christoffersen_independence(clustered);
    const ChristoffersenResult s = christoffersen_independence(spread);
    EXPECT_LT(c.pvalue, 0.001) << "a run of 10 exceptions must reject independence";
    EXPECT_GT(c.pi11, c.pi01);  // exception begets exception in the cluster
    EXPECT_GT(s.pvalue, 0.10) << "evenly spread exceptions must not reject";
    EXPECT_GT(c.lr, s.lr);
    EXPECT_THROW(christoffersen_independence(std::vector<std::uint8_t>{1}),
                 std::invalid_argument);
}

TEST(Christoffersen, ConditionalCoverageIsSumOfComponents) {
    std::vector<std::uint8_t> ex(250, 0);
    for (int t = 100; t < 108; ++t) ex[t] = 1;
    const KupiecResult uc = kupiec_pof(250, 8, 0.01);
    const ChristoffersenResult ind = christoffersen_independence(ex);
    const ConditionalCoverageResult cc = christoffersen_cc(ex, 0.01);
    EXPECT_NEAR(cc.lr, uc.lr + ind.lr, 1e-12);
    EXPECT_NEAR(cc.lr_uc, uc.lr, 1e-12);
    EXPECT_NEAR(cc.lr_ind, ind.lr, 1e-12);
    EXPECT_NEAR(cc.pvalue, chi2_sf(cc.lr, 2.0), 1e-14);
    EXPECT_LT(cc.pvalue, 0.05);  // wrong rate AND clustered: joint rejection
}

TEST(Basel, ExactZoneBoundariesAt250Obs) {
    // Green: 0-4 exceptions, multiplier 3.0.
    EXPECT_EQ(basel_traffic_light(0).zone, BaselZone::Green);
    EXPECT_EQ(basel_traffic_light(4).zone, BaselZone::Green);
    EXPECT_DOUBLE_EQ(basel_traffic_light(4).multiplier, 3.0);
    // Yellow: 5-9 with the regulatory add-on ladder.
    EXPECT_EQ(basel_traffic_light(5).zone, BaselZone::Yellow);
    EXPECT_DOUBLE_EQ(basel_traffic_light(5).multiplier, 3.40);
    EXPECT_DOUBLE_EQ(basel_traffic_light(6).multiplier, 3.50);
    EXPECT_DOUBLE_EQ(basel_traffic_light(7).multiplier, 3.65);
    EXPECT_DOUBLE_EQ(basel_traffic_light(8).multiplier, 3.75);
    EXPECT_EQ(basel_traffic_light(9).zone, BaselZone::Yellow);
    EXPECT_DOUBLE_EQ(basel_traffic_light(9).multiplier, 3.85);
    // Red: 10 or more, multiplier 4.0.
    EXPECT_EQ(basel_traffic_light(10).zone, BaselZone::Red);
    EXPECT_DOUBLE_EQ(basel_traffic_light(10).multiplier, 4.0);
    EXPECT_EQ(basel_traffic_light(15).zone, BaselZone::Red);
    EXPECT_THROW(basel_traffic_light(-1), std::invalid_argument);
}

TEST(Basel, BinomialZoneProbabilities) {
    // Under a correct 99 % model the green zone covers ~89.2 % of outcomes
    // and 9 exceptions is already the 99.97th percentile (scipy reference).
    EXPECT_NEAR(basel_traffic_light(4).cumulative_prob, 0.89218762690362508, 1e-10);
    EXPECT_NEAR(basel_traffic_light(9).cumulative_prob, 0.99974980993125950, 1e-10);
    EXPECT_EQ(to_string(BaselZone::Green), "green");
    EXPECT_EQ(to_string(BaselZone::Yellow), "yellow");
    EXPECT_EQ(to_string(BaselZone::Red), "red");
}

File: rust/equity-var-engine/tests/test_backtest.rs
//! Backtesting: Kupiec LR hand-computed, chi2 p-values, Christoffersen
//! clustering detection, Basel traffic-light exact zone boundaries.

use eq_var_engine::backtest::*;
use eq_var_engine::stats::chi2_sf;

#[test]
fn exceptions_indicator_and_broadcast() {
    let pnl = [-120.0, 50.0, -99.9, -100.1, 0.0];
    let var1 = [100.0]; // scalar broadcast
    let ex = exceptions_from_pnl(&pnl, &var1).unwrap();
    assert_eq!(ex, vec![1, 0, 0, 1, 0]);
    let vard = [130.0, 100.0, 90.0, 100.0, 100.0];
    assert_eq!(exceptions_from_pnl(&pnl, &vard).unwrap(), vec![0, 0, 1, 1, 0]);
    assert!(exceptions_from_pnl(&pnl, &[-1.0]).is_err());
    assert!(exceptions_from_pnl(&pnl, &[1.0, 2.0]).is_err());
}

#[test]
fn kupiec_hand_computed_lr() {
    let (t, x, p): (f64, f64, f64) = (250.0, 5.0, 0.01);
    let pihat = x / t;
    let ll0 = (t - x) * (1.0 - p).ln() + x * p.ln();
    let ll1 = (t - x) * (1.0 - pihat).ln() + x * pihat.ln();
    let lr_hand = -2.0 * (ll0 - ll1);
    let r = kupiec_pof(250, 5, 0.01).unwrap();
    assert!((r.lr - lr_hand).abs() < 1e-12);
    assert!((r.pvalue - chi2_sf(lr_hand, 1.0).unwrap()).abs() < 1e-14);
    assert!((r.expected - 2.5).abs() < 1e-12);
    assert!((r.rate - 0.02).abs() < 1e-12);
}

#[test]
fn kupiec_chi2_critical_value_gives_point_oh_five() {
    assert!((chi2_sf(3.841458820694124, 1.0).unwrap() - 0.05).abs() < 1e-4);
    let exact = kupiec_pof(200, 2, 0.01).unwrap();
    assert!(exact.lr.abs() < 1e-12);
    assert!((exact.pvalue - 1.0).abs() < 1e-12);
}

#[test]
fn kupiec_degenerate_counts_and_monotonicity() {
    let zero = kupiec_pof(250, 0, 0.01).unwrap();
    assert!((zero.lr - (-2.0 * 250.0 * 0.99f64.ln())).abs() < 1e-12);
    assert!(kupiec_pof(250, 10, 0.01).unwrap().lr > kupiec_pof(250, 6, 0.01).unwrap().lr);
    assert!(kupiec_pof(250, 10, 0.01).unwrap().pvalue < 0.05);
    assert!(kupiec_pof(0, 0, 0.01).is_err());
    assert!(kupiec_pof(250, 251, 0.01).is_err());
    assert!(kupiec_pof(250, -1, 0.01).is_err());
}

#[test]
fn christoffersen_transition_counts_on_tiny_pattern() {
    let ex = [0u8, 1, 1, 0, 1];
    let r = christoffersen_independence(&ex).unwrap();
    assert!((r.n00 - 0.0).abs() < 1e-12);
    assert!((r.n01 - 2.0).abs() < 1e-12);
    assert!((r.n10 - 1.0).abs() < 1e-12);
    assert!((r.n11 - 1.0).abs() < 1e-12);
    assert!((r.pi01 - 1.0).abs() < 1e-12);
    assert!((r.pi11 - 0.5).abs() < 1e-12);
}

#[test]
fn christoffersen_detects_planted_clustering() {
    let mut clustered = vec![0u8; 250];
    let mut spread = vec![0u8; 250];
    for v in clustered.iter_mut().take(110).skip(100) {
        *v = 1;
    }
    let mut t = 12;
    while t < 250 {
        spread[t] = 1;
        t += 25;
    }
    let c = christoffersen_independence(&clustered).unwrap();
    let s = christoffersen_independence(&spread).unwrap();
    assert!(c.pvalue < 0.001, "a run of 10 exceptions must reject independence");
    assert!(c.pi11 > c.pi01);
    assert!(s.pvalue > 0.10, "evenly spread exceptions must not reject");
    assert!(c.lr > s.lr);
    assert!(christoffersen_independence(&[1u8]).is_err());
}

#[test]
fn christoffersen_conditional_coverage_is_sum_of_components() {
    let mut ex = vec![0u8; 250];
    for v in ex.iter_mut().take(108).skip(100) {
        *v = 1;
    }
    let uc = kupiec_pof(250, 8, 0.01).unwrap();
    let ind = christoffersen_independence(&ex).unwrap();
    let cc = christoffersen_cc(&ex, 0.01).unwrap();
    assert!((cc.lr - (uc.lr + ind.lr)).abs() < 1e-12);
    assert!((cc.lr_uc - uc.lr).abs() < 1e-12);
    assert!((cc.lr_ind - ind.lr).abs() < 1e-12);
    assert!((cc.pvalue - chi2_sf(cc.lr, 2.0).unwrap()).abs() < 1e-14);
    assert!(cc.pvalue < 0.05);
}

#[test]
fn basel_exact_zone_boundaries_at_250_obs() {
    assert_eq!(basel_traffic_light(0, 250).unwrap().zone, BaselZone::Green);
    let r4 = basel_traffic_light(4, 250).unwrap();
    assert_eq!(r4.zone, BaselZone::Green);
    assert!((r4.multiplier - 3.0).abs() < 1e-12);

    let r5 = basel_traffic_light(5, 250).unwrap();
    assert_eq!(r5.zone, BaselZone::Yellow);
    assert!((r5.multiplier - 3.40).abs() < 1e-12);
    assert!((basel_traffic_light(6, 250).unwrap().multiplier - 3.50).abs() < 1e-12);
    assert!((basel_traffic_light(7, 250).unwrap().multiplier - 3.65).abs() < 1e-12);
    assert!((basel_traffic_light(8, 250).unwrap().multiplier - 3.75).abs() < 1e-12);
    let r9 = basel_traffic_light(9, 250).unwrap();
    assert_eq!(r9.zone, BaselZone::Yellow);
    assert!((r9.multiplier - 3.85).abs() < 1e-12);

    let r10 = basel_traffic_light(10, 250).unwrap();
    assert_eq!(r10.zone, BaselZone::Red);
    assert!((r10.multiplier - 4.0).abs() < 1e-12);
    assert_eq!(basel_traffic_light(15, 250).unwrap().zone, BaselZone::Red);
    assert!(basel_traffic_light(-1, 250).is_err());
}

#[test]
fn basel_binomial_zone_probabilities() {
    assert!(
        (basel_traffic_light(4, 250).unwrap().cumulative_prob - 0.892_187_626_903_625_1).abs()
            < 1e-10
    );
    assert!(
        (basel_traffic_light(9, 250).unwrap().cumulative_prob - 0.999_749_809_931_259_5).abs()
            < 1e-10
    );
    assert_eq!(BaselZone::Green.as_str(), "green");
    assert_eq!(BaselZone::Yellow.as_str(), "yellow");
    assert_eq!(BaselZone::Red.as_str(), "red");
}
